### DISCUSSION TITLE
How to load a CMYK image, convert it to RGB and strip data?

### DISCUSSION THREAD

**Discussion by mlocati** [ImageMagick/ImageMagick#4387] [Help]
I'm maintaining a PHP library that abstracts different graphics engines (gd, graphicsmagick and imagemagick).

I can't find a way to correctly load a CMYK image, converting it to RGB and stripping the metadata.

Here's a sample code that mimics what's done by the PHP library:

```c
// compile with:
// gcc $(MagickWand-config --cflags) -o test test.c $(MagickWand-config --ldflags --libs) -lm

#include <stdio.h>
#include <math.h>
#include <MagickWand/MagickWand.h>

char* readRGBProfile(size_t* length)
{
    char* result = NULL;
    FILE* fd = fopen("sRGB_IEC61966-2-1_black_scaled.icc", "rb");
    if (fd == NULL) {
        printf("Failed to read the ICC profile\n");
    } else {
        fseek(fd, 0, SEEK_END);
        *length = ftell(fd);
        result = (char*) malloc(*length);
        if (result == NULL) {
            printf("malloc() failed\n");
        } else {
            fseek(fd, 0, SEEK_SET);
            if (fread(result, 1, *length, fd) != *length) {
                printf("fread() failed\n");
                free(result);
                result = NULL;
            }
        }
        fclose(fd);
    }
    if (result == NULL) {
        *length = 0;
    }
    return result;
}

void ExitWithError(MagickWand* mw)
{
    ExceptionType severity;
    char* description = MagickGetException(mw, &severity);
    if (description != NULL && description[0] != '\0')
    {
        printf("%s\n", description);
    }
    else
    {
        printf("failed (unknown reason)\n");
    }
    DestroyMagickWand(mw);
    MagickWandTerminus();
    exit(1);
}

int main(int argc, char** argv)
{
    MagickWand* frame;
    PixelWand* pixel;

    MagickWandGenesis();

    size_t imageMagickVersion;
    printf("ImageMagick version: %s\n", MagickGetVersion(&imageMagickVersion));

    // $rgbProfile = file_get_contents('sRGB_IEC61966-2-1_black_scaled.icc');
    printf("Reading profile file... ");
    size_t rgbProfileLength;
    char* rgbProfile = readRGBProfile(&rgbProfileLength);
    if (rgbProfile == NULL) {
        MagickWandTerminus();
        return 1;
    }
    printf("done.\n");

    // $image = new Imagick('pixel-CMYK.jpg');
    printf("Opening image... ");
    MagickWand* image = NewMagickWand();
    if (MagickReadImage(image, "pixel-CMYK.jpg") != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    if (MagickSetImageFilename(image,  "pixel-CMYK.jpg") != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    MagickSetLastIterator(image);
    printf("done.\n");

    // $image->setType(Imagick::IMGTYPE_TRUECOLORALPHA);
    printf("Setting image type... ");
    if (MagickSetType(image, TrueColorAlphaType) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // foreach ($image as $frame) {
    printf("Setting frames alpha channel... ");
    if (MagickSetIteratorIndex(image, 0) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    frame = MagickGetImage(image);
    if (frame == NULL) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //$frame->setimagealphachannel(Imagick::ALPHACHANNEL_REMOVE);
    if (MagickSetImageAlphaChannel(frame, RemoveAlphaChannel) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //}
    printf("done.\n");

    // $image->setColorspace(Imagick::COLORSPACE_CMYK);
    printf("Setting image colorspace... ");
    if (MagickSetColorspace(image, CMYKColorspace) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $image->profileImage('icc', $rgbProfile);
    printf("Setting image profile... ");
    if (MagickProfileImage(image, "icc", rgbProfile, rgbProfileLength) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $image->setType(Imagick::IMGTYPE_TRUECOLORALPHA);
    printf("Setting image type... ");
    if (MagickSetType(image, TrueColorAlphaType) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    //foreach ($image as $frame) {
    printf("Setting frames colorspace and alpha channel... ");
    if (MagickSetIteratorIndex(image, 0) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    frame = MagickGetImage(image);
    if (frame == NULL) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //$frame->transformimagecolorspace(Imagick::COLORSPACE_RGB);
    if (MagickTransformImageColorspace(frame, RGBColorspace) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //$frame->setimagealphachannel(Imagick::ALPHACHANNEL_SET);
    if (MagickSetImageAlphaChannel(frame, SetAlphaChannel) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //}
    printf("done.\n");

    // $image->transformimagecolorspace(Imagick::COLORSPACE_RGB);
    printf("Setting image colorspace... ");
    if (MagickTransformImageColorspace(image, RGBColorspace) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    //$image->setColorspace(Imagick::COLORSPACE_RGB);
    if (MagickSetColorspace(image, RGBColorspace) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $image->profileImage('icc', $rgbProfile);
    printf("Setting image profile... ");
    if (MagickProfileImage(image, "icc", rgbProfile, rgbProfileLength) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $image->stripImage();
    printf("Stripping image... ");
    if (MagickStripImage(image) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $pixel = $image->getImagePixelColor(0, 0);
    printf("Getting pixel... ");
    pixel = NewPixelWand();
    if (MagickGetImagePixelColor(image, 0, 0 , pixel) != MagickTrue) {
        free(rgbProfile);
        ExitWithError(image);
    }
    printf("done.\n");

    // $alpha = round($pixel->getColorValue(Imagick::COLOR_ALPHA) * 100));
    int alpha = (int) round(PixelGetAlpha(pixel) * 100);
    printf("Alpha: expected %d, got %d\n", 100, alpha);
    // $red = round($pixel->getColorValue(Imagick::COLOR_RED) * 255));
    int red = (int) round(PixelGetRed(pixel) * 255);
    printf("Red  : expected %d, got %d\n", 0, red);
    // $green = round($pixel->getColorValue(Imagick::COLOR_GREEN) * 255));
    int green = (int) round(PixelGetGreen(pixel) * 255);
    printf("Green: expected %d, got %d\n", 130, green);
    // $blue = round($pixel->getColorValue(Imagick::COLOR_BLUE) * 255));
    int blue = (int) round(PixelGetBlue(pixel) * 255);
    printf("Blue : expected %d, got %d\n", 162, blue);

    DestroyMagickWand(image);
    MagickWandTerminus();
    return 0;
}
```

(Everything is included [in this archive](https://github.com/ImageMagick/ImageMagick/files/7371999/test.zip)).

The conversion fails:

```
ImageMagick version: ImageMagick 7.1.0-8 Q16-HDRI x86_64 2021-09-18 https://imagemagick.org
Reading profile file... done.
Opening image... done.
Setting image type... done.
Setting frames alpha channel... done.
Setting image colorspace... done.
Setting image profile... done.
Setting image type... done.
Setting frames colorspace and alpha channel... done.
Setting image colorspace... done.
Setting image profile... done.
Stripping image... done.
Getting pixel... done.
Alpha: expected 100, got 100
Red  : expected 0, got 0
Green: expected 130, got 58
Blue : expected 162, got 93
```

instead of this color:
![expected](https://user-images.githubusercontent.com/928116/137881736-87ba84cd-da42-4785-8ea8-d7b65141f73e.png)
I have this color:
![actual](https://user-images.githubusercontent.com/928116/137881891-9316b061-7424-4061-a76b-b310b0393588.png)

**Comment by snibgo** [comment]
You do many operations. At the end, the output is not what you expected.

So you must know what you expect after each single operation. So you can test whether the values are what you expect after each single operation.

After a brief look at your operations, I notice that they are strange. You read an CMYK image, with embedded profile, and "MagickSetType(image, TrueColorAlphaType)" which tells IM that the image isn't really CMYK, but RGBA.

Your code uses both MagickSetColorspace and MagickProfileImage on the same image. This is a bad idea.

I have no idea what your code is supposed to do, or supposed to test.

**Reply by mlocati** [reply]
We have so many operations because the library doesn't simply strip images. Images can come from files or built on the fly, and images may have different colorspaces and profiles.
So, the steps above are all the ones performed by higher-lever functions:

- It first loads the image (the operations from "Opening image" to the first "Setting image colorspace")
- Then it changes the colorspace to RGB (the operations from the first "Setting image profile" to the second "Setting image colorspace")
- Then it strips the image (from the second "Setting image profile" to "Stripping image")

**Comment by snibgo** [comment]
You start with a CMYK image with embedded profile. If you want to convert that to sRGB, use MagickProfileImage() with your desired ICC profile file. That is all you need to do, and all you should do. Nothing else. Don't also change the type or the colorspace.

**Reply by mlocati** [reply]
It seems I can't even load the CMYK correctly... Am I missing something?

```c
// compile with:
// gcc $(MagickWand-config --cflags) -o test test.c $(MagickWand-config --ldflags --libs) -lm -Wall

#include <stdio.h>
#include <math.h>
#include <MagickWand/MagickWand.h>

#define EXPECTED_ALPHA 100

#define EXPECTED_RED 0x00
#define EXPECTED_GREEN 0x82
#define EXPECTED_BLUE 0xA2

#define EXPECTED_CYAN 100
#define EXPECTED_MAGENTA 20
#define EXPECTED_YELLOW 0
#define EXPECTED_BLACK 36

void ExitWithError(MagickWand* image)
{
    ExceptionType severity;
    char* description = MagickGetException(image, &severity);
    if (description != NULL && description[0] != '\0')
    {
        printf("%s\n", description);
    }
    else
    {
        printf("failed (unknown reason)\n");
    }
    DestroyMagickWand(image);
    MagickWandTerminus();
    exit(1);
}

void CheckPixelRGB(MagickWand* image)
{
    printf("Getting RGB pixel... ");
    PixelWand* pixel = NewPixelWand();
    if (MagickGetImagePixelColor(image, 0, 0 , pixel) != MagickTrue) {
        ExitWithError(image);
    }
    int alpha = (int) round(PixelGetAlpha(pixel) * 100);
    int red = (int) round(PixelGetRed(pixel) * 255);
    int green = (int) round(PixelGetGreen(pixel) * 255);
    int blue = (int) round(PixelGetBlue(pixel) * 255);
    DestroyPixelWand(pixel);
    if (alpha != EXPECTED_ALPHA || red != EXPECTED_RED || green != EXPECTED_GREEN || blue != EXPECTED_BLUE) {
        printf("WRONG!\n");
        printf("- Alpha: expected %d%%, got %d%%\n", EXPECTED_ALPHA, alpha);
        printf("- Red  : expected 0x%02X, got 0x%02X\n", EXPECTED_RED, red);
        printf("- Green: expected 0x%02X, got 0x%02X\n", EXPECTED_GREEN, green);
        printf("- Blue : expected 0x%02X, got 0x%02X\n", EXPECTED_BLUE, blue);
    } else {
        printf("Pixel is ok\n");
    }
}

void CheckPixelCMYK(MagickWand* image)
{
    printf("Getting CMYK pixel... ");
    PixelWand* pixel = NewPixelWand();
    if (MagickGetImagePixelColor(image, 0, 0 , pixel) != MagickTrue) {
        ExitWithError(image);
    }
    int alpha = (int) round(PixelGetAlpha(pixel) * 100);
    int cyan = (int) round(PixelGetCyan(pixel) * 100);
    int magenta = (int) round(PixelGetMagenta(pixel) * 100);
    int yellow = (int) round(PixelGetYellow(pixel) * 100);
    int black = (int) round(PixelGetBlack(pixel) * 100);
    DestroyPixelWand(pixel);
    if (alpha != EXPECTED_ALPHA || cyan != EXPECTED_CYAN || magenta != EXPECTED_MAGENTA || yellow != EXPECTED_YELLOW || black != EXPECTED_BLACK) {
        printf("WRONG!\n");
        printf("- Alpha  : expected %d%%, got %d%%\n", EXPECTED_ALPHA, alpha);
        printf("- Cyan   : expected %d%%, got %d%%\n", EXPECTED_CYAN, cyan);
        printf("- Magenta: expected %d%%, got %d%%\n", EXPECTED_MAGENTA, magenta);
        printf("- Yellow : expected %d%%, got %d%%\n", EXPECTED_YELLOW, yellow);
        printf("- Black  : expected %d%%, got %d%%\n", EXPECTED_BLACK, black);
    } else {
        printf("Pixel is ok\n");
    }
}

int main(int argc, char** argv)
{
    MagickWandGenesis();

    size_t imageMagickVersion;
    printf("ImageMagick version: %s\n", MagickGetVersion(&imageMagickVersion));

    printf("Opening image... ");
    MagickWand* image = NewMagickWand();
    if (MagickReadImage(image, "pixel-CMYK.jpg") != MagickTrue) {
        ExitWithError(image);
    }
    printf("done.\n");
    CheckPixelCMYK(image);
    CheckPixelRGB(image);

    DestroyMagickWand(image);
    MagickWandTerminus();
    return 0;
}
```

Output: 

```
ImageMagick version: ImageMagick 7.1.0-8 Q16-HDRI x86_64 2021-09-18 https://imagemagick.org
Opening image... done.
Getting CMYK pixel... WRONG!
- Alpha  : expected 100%, got 100%
- Cyan   : expected 100%, got 99%
- Magenta: expected 20%, got 0%
- Yellow : expected 0%, got 31%
- Black  : expected 36%, got 23%
Getting RGB pixel... WRONG!
- Alpha: expected 100%, got 100%
- Red  : expected 0x00, got 0xFC
- Green: expected 0x82, got 0x01
- Blue : expected 0xA2, got 0x4E
```

PS: the `pixel-CMYK.jpg` file is the one included in the [original message](https://github.com/ImageMagick/ImageMagick/discussions/4387#discussion-3636018)
PPS: the expected colors are the ones that I see when viewing the image with Gimp

**Comment by snibgo** [comment]
You have:
```
#define EXPECTED_CYAN 100
#define EXPECTED_MAGENTA 20
#define EXPECTED_YELLOW 0
#define EXPECTED_BLACK 36
```
You say those expected values came from Gimp. From the Color Picker? Gimp processes in RGB, so it has converted your image from CMYK to RGB, and converted back to CMYK for the color picker. So these may not be the same as the actual CMYK values from the file. "magick pixel-CMYK.jpg -verbose info:" tells us what the values are:
```
    Cyan:
      min: 252  (0.988235)
      max: 252 (0.988235)
      mean: 252 (0.988235)
      median: 252 (0.988235)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
    Magenta:
      min: 1  (0.00392157)
      max: 1 (0.00392157)
      mean: 1 (0.00392157)
      median: 1 (0.00392157)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
    Yellow:
      min: 78  (0.305882)
      max: 78 (0.305882)
      mean: 78 (0.305882)
      median: 78 (0.305882)
      standard deviation: 0 (0)
      kurtosis: 6.4e+49
      skewness: 0
      entropy: 0
    Black:
      min: 59  (0.231373)
      max: 59 (0.231373)
```
For example, cyan has min=max=252 on a scale of 0 to 255, which is 0.988235 on a scale of 0 to 1, so 98.8%. This is close to your EXPECTED_CYAN, but the other values are very different.

The values you actually got are exactly correct (to the nearest percent).

**Reply by mlocati** [reply]
That's really strange... as you say, `magick pixel-CMYK.jpg -verbose info:` returns these values

C: 99%
M: 0%
Y: 31%
K: 23%

here's an image I created with Gimp with those values:

![immagine](https://user-images.githubusercontent.com/928116/137949474-d54340e6-7c41-4c5f-af48-4e0855750919.png)

but when I open the image with Gimp, I have this color (picked with the color picker):

![immagine](https://user-images.githubusercontent.com/928116/137949630-0eed1e49-9d28-45ca-9151-5084147dcd91.png)

They are very different... who's wrong? Gimp or ImageMagick?

**Reply by mlocati** [reply]
I also have the light blue color when I run `magick pixel-RGB.jpg -colorspace sRGB converted.png` and then I open the `converted.png` image with Gimp.

PS: running `magick converted.png -verbose info:` gives
```
Class: PseudoClass
  Geometry: 1x1+0+0
  Resolution: 300x300
  Print size: 0.00333333x0.00333333
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: Palette
```

with these RGB values:
R 0x30
G: 0x82
B: 0xA0

**Comment by snibgo** [comment]
> They are very different... who's wrong? Gimp or ImageMagick?

IM is reporting the CMYK values that are in the JPG file. Instead of using "-verbose info:" we can:
```
magick pixel-CMYK.jpg txt:

# ImageMagick pixel enumeration: 1,1,255,cmyk
0,0: (252,1,78,59)  #FC014E3B  cmyk(252,1,78,59)
```
As implied by the header, these values are on a scale of 0 to 255.

Gimp is reporting something different. It is reporting the result of converting the CMYK colour to sRGB (how? I don't know), then converting that sRGB colour back to CMYK (how? I don't know).

**Reply by mlocati** [reply]
cmyk(252,1,78,59) is a light green color:
- with https://www.w3schools.com/colors/colors_cmyk.asp
   ![immagine](https://user-images.githubusercontent.com/928116/137955577-e222a5e2-6eab-406d-a864-843cf698717b.png)
- with Gimp
   ![immagine](https://user-images.githubusercontent.com/928116/137956000-35d2e440-8c9f-4f49-9b27-6a1b71d6bf09.png)

But opening the image with Gimp and Krita and zooming in, the color is very different:
- Gimp: 
   ![immagine](https://user-images.githubusercontent.com/928116/137957388-85e394f4-7bb5-4b04-903e-ed25cadab692.png)
- Krita:
   ![immagine](https://user-images.githubusercontent.com/928116/137958120-6f755375-6f70-491b-853f-ecdd1e28d5cb.png)

And that's really strange, since for example Krita seems to recognize the profile correctly:
![immagine](https://user-images.githubusercontent.com/928116/137958591-d0e3e9c6-3a69-4077-aa63-c462d09312f8.png)

**Comment by fmw42** [comment]
If the image does not have a proper CMYK profile, many tools do not interpret CMYK files properly. So add a profile such as USWebCoatedSWOP.icc and see what happens.

**Comment by mlocati** [comment]
I think the icc profile of the jpg image is correct (extracting it with MagickGetImageProfile and loading it back with MagickSetImageProfile doesn't thow any error).
It'd be interesting to see how photoshop behaves with this jpg...

**Comment by snibgo** [comment]
The file pixel-CMYK.jpg is pixel-CMYK.jpg is encoded as CMYK and contains a valid CMYK ICC profile.

**Reply by mlocati** [reply]
I asked a friend to open the image with Photoshop.
When opening the image, Photoshop asks the following:
> The document "pixel-CMYK.jpg" contains an embedded color profile that doesn't match to the current CMYK workspace.
> Embedded: Eurostandard (couché), 9%, GCR, Maxim...
> Workspace: SWOP (coated), 20%, GCR, Media
> What do you want to do?
> 1. Use the embedded profile
> 2. Convert colors to workspace
> 3. Delete embedded colorspace

When opening the image with 1, we have this color:
![embedded.png](https://user-images.githubusercontent.com/928116/138057914-95139d36-b9e0-47d6-b987-d2fd386a0340.png)

When opening the image with 2, we have this color:
![converted.png](https://user-images.githubusercontent.com/928116/138057959-5460fdbd-7a55-40a5-bae3-14ff812e26d3.png)

When opening the image with 3, we have this color:
![deleted.png](https://user-images.githubusercontent.com/928116/138058025-54afffad-fca7-4000-8e8e-7561d44a763d.png)

So, to sum everything up, with Gimp, Krita and Photoshop we see a light blue color, whereas ImageMagick says it's a light green color.
Something strange is happening here...

**Comment by snibgo** [comment]
> ... whereas ImageMagick says it's a light green color.

What does that mean? What IM process or command says pixel-CMYK.jpg is a light green colour?

**Comment by fmw42** [comment]
I get:

`convert -size 100x100 xc:"cmyk(252,1,78,59)" swatch.jpg`

![swatch](https://user-images.githubusercontent.com/13458218/138150261-aa5d0569-d0cf-4401-bbe4-bdbd5af1f303.jpg)

`convert -size 100x100 xc:"cmyk(252,1,78,59)" -profile /Users/fred/Images/Profiles/WebCoatedFOGRA28.icc swatch2.jpg`

![swatch2](https://user-images.githubusercontent.com/13458218/138150302-0615dbaf-45d1-4d2d-8a2a-af6b40d5a9f8.jpg)

IM 6.9.12-25 Q16 x86_64 Mac OSX Sierra

**Comment by snibgo** [comment]
When I view fmw42's post in Firefox, the two images seem to be greenish.

When I download the images and view them in Windows Photo Viewer, they seem to be blueish.

Viewing the downloaded images in Gimp, the first (with no embedded profile) is greenish. The second (with embedded profile) seems blueish.

This demonstrates that those programs convert CMYK images to RGB differently. I don't know why, or which is correct.

But this says nothing about ImageMagick. Is IM doing anything wrong?
